README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->
# Color Tools
![Flutter CI](https://github.com/Ocell-io/color-tools/actions/workflows/main.yml/badge.svg)  
Color Tools is a library that provies ulitlity functions for the flutter Color class. In comparison to other libaries it tries to avoid adding additonal classes where possible.

## Features

* create Color from names
* create Color from color codes
* convert Colors to various types of Strings
* lighten a color
* darken a color
* saturate a color
* desaturate a color
* and more


## Getting started
color_tools provides options to load colors via ColorFrom and add various extensions to the Color class to make it more user friendly.


```dart
import 'package:color_tools/color_tools.dart';

final code = ColorFrom.name('red').toCode();
assert(code == '#ff0000')
```

## Usage

Create a color from a name
```dart
final color = ColorFrom.name('red');
```

Create a color from a color code
```dart
final color = ColorFrom.code('#ff0000');
```

Generate a specific color from an object. Useful for creating colors for other objects (usernames).
```dart
final someObject = "does not have to be a string!";
final color = ColorFrom.object(someObject);

//called for the same object it always gives the same result
assert(ColorFrom.object(someObject) == ColorFrom.object(someObject)) ;

//you can limit the range of possible colors
ColorFrom.object(someObject,
    hueLowerLimit: 40, valueLowerLimit: 1, saturationUpperLimit: 1),

```

Lighten a color
```dart
final color = Colors.red.lighten();
```

Saturate a color
```dart
final color = Colors.red.saturate();
```


CHANGELOG.md
## 0.1.0

* initial release :)


test/util.dart
import 'package:flutter/painting.dart';

const red = Color(0xffff0000);
const redCode = "#ff0000";
const redCodeAlpha = "#ffff0000";
const redPurpleRGB = "(255, 0, 0)";
const redPurpleARGB = "(255, 255, 0, 0)";

const darkPurple = Color(0xDD482191);
const darkPurpleCode = "#482191";
const darkPurpleCodeWithAlpha = "#dd482191";
const darkPurpleRGB = "(72, 33, 145)";
const darkPurpleARGB = "(221, 72, 33, 145)";


test/color_to_test.dart
import 'package:color_tools/color_tools.dart';
import 'package:flutter_test/flutter_test.dart';

import 'util.dart';

void main() {
  test('test color to code', () {
    expect(red.toCode(), redCode);
    expect(darkPurple.toCode(), darkPurpleCode);
  });

  test('test color to ARGB', () {
    expect(red.toARGB(), redPurpleARGB);
    expect(darkPurple.toARGB(), darkPurpleARGB);
  });
  test('test color to RGB', () {
    expect(red.toRGB(), redPurpleRGB);
    expect(darkPurple.toRGB(), darkPurpleRGB);
  });

  test('test color to name', () {
    expect(red.toName(), 'red');
  });

  test('test color roundtrip', () {
    expect(ColorFrom.name('red')!.toName(), 'red');
  });

  test('test color to toCodeWithAlpha', () {
    expect(darkPurple.toCodeWithAlpha(), darkPurpleCodeWithAlpha);
  });
}


test/color_from_test.dart
import 'package:color_tools/color_tools.dart';
import 'package:flutter_test/flutter_test.dart';

import 'util.dart';

void main() {
  test('test color from name', () {
    expect(ColorFrom.name("red"), equals(red));
  });

  test('test color from code w alpha', () {
    expect(ColorFrom.code(redCodeAlpha), equals(red));
  });

  test('test color from code', () {
    expect(ColorFrom.code(redCode), equals(red));
  });

  test('test color from noAlpha', () {
    expect(ColorFrom.noAlpha(0xff0000), equals(red));
  });

  test('test color from noAlpha', () {
    expect(ColorFrom.string("red"), equals(red));
    expect(ColorFrom.string(redCode), equals(red));
  });

  test('test color from object', () {
    const someObject = "does not have to be a string!";

    expect(ColorFrom.object(someObject), equals(ColorFrom.object(someObject)));
    expect(
        ColorFrom.object(someObject,
            valueLowerLimit: 1, saturationUpperLimit: 1),
        equals(ColorFrom.object(someObject,
            valueLowerLimit: 1, saturationUpperLimit: 1)));
    expect(
        ColorFrom.object(someObject,
            hueLowerLimit: 40, valueLowerLimit: 1, saturationUpperLimit: 1),
        equals(ColorFrom.object(someObject,
            hueLowerLimit: 40, valueLowerLimit: 1, saturationUpperLimit: 1)));
  });
}


lib/color_to.dart
import 'package:color_tools/color_tabels.dart';
import 'package:flutter/painting.dart';

extension ColorTo on Color {
  /// Converts a color to a color code #rrggbb
  ///
  /// ```dart
  /// Color(0xFFAABBCC).toColorCode() == '#aabbcc'
  /// ```
  String toCode() {
    final r = red.toRadixString(16).padLeft(2, '0');
    final g = green.toRadixString(16).padLeft(2, '0');
    final b = blue.toRadixString(16).padLeft(2, '0');
    return '#$r$g$b';
  }

  /// Converts a color to a color code #aarrggbb
  ///
  /// ```dart
  /// Color(0xFFAABBCC).toCodeWithAlpha() == '#ffaabbcc'
  /// ```
  String toCodeWithAlpha() {
    final a = alpha.toRadixString(16).padLeft(2, '0');
    final r = red.toRadixString(16).padLeft(2, '0');
    final g = green.toRadixString(16).padLeft(2, '0');
    final b = blue.toRadixString(16).padLeft(2, '0');
    return '#$a$r$g$b';
  }

  /// Converts a color to a rgb string
  ///
  /// ```dart
  /// Color(0xFFAABBCC).toRGB() == '(170, 187, 204)'
  /// ```
  String toRGB() {
    return '($red, $green, $blue)';
  }

  /// Converts a color to a alpha + rgb string
  ///
  /// ```dart
  /// Color(0xFFAABBCC).toARGB() == '(255, 170, 187, 204)'
  /// ```
  String toARGB() {
    return '($alpha, $red, $green, $blue)';
  }

  /// Converts a color to a color name
  ///
  /// Note that this is agnostic to alpha
  ///
  /// ```dart
  /// Color(0xFF000000).toName() == 'white'
  /// ```
  String? toName() {
    return colorValueToName[value | 0xFF000000];
  }
}


lib/color_tabels.dart
/// Color name to value from
/// http://www.w3.org/TR/css3-color/#svg-color
const colorNameToValue = {
  "aliceblue": 0xFFF0F8FF,
  "antiquewhite": 0xFFFAEBD7,
  "aqua": 0xFF00FFFF,
  "aquamarine": 0xFF7FFFD4,
  "azure": 0xFFF0FFFF,
  "beige": 0xFFF5F5DC,
  "bisque": 0xFFFFE4C4,
  "black": 0xFF000000,
  "blanchedalmond": 0xFFFFEBCD,
  "blue": 0xFF0000FF,
  "blueviolet": 0xFF8A2BE2,
  "brown": 0xFFA52A2A,
  "burlywood": 0xFFDEB887,
  "cadetblue": 0xFF5F9EA0,
  "chartreuse": 0xFF7FFF00,
  "chocolate": 0xFFD2691E,
  "coral": 0xFFFF7F50,
  "cornflowerblue": 0xFF6495ED,
  "cornsilk": 0xFFFFF8DC,
  "crimson": 0xFFDC143C,
  "cyan": 0xFF00FFFF,
  "darkblue": 0xFF00008B,
  "darkcyan": 0xFF008B8B,
  "darkgoldenrod": 0xFFB8860B,
  "darkgray": 0xFFA9A9A9,
  "darkgreen": 0xFF006400,
  "darkgrey": 0xFFA9A9A9,
  "darkkhaki": 0xFFBDB76B,
  "darkmagenta": 0xFF8B008B,
  "darkolivegreen": 0xFF556B2F,
  "darkorange": 0xFFFF8C00,
  "darkorchid": 0xFF9932CC,
  "darkred": 0xFF8B0000,
  "darksalmon": 0xFFE9967A,
  "darkseagreen": 0xFF8FBC8F,
  "darkslateblue": 0xFF483D8B,
  "darkslategray": 0xFF2F4F4F,
  "darkslategrey": 0xFF2F4F4F,
  "darkturquoise": 0xFF00CED1,
  "darkviolet": 0xFF9400D3,
  "deeppink": 0xFFFF1493,
  "deepskyblue": 0xFF00BFFF,
  "dimgray": 0xFF696969,
  "dimgrey": 0xFF696969,
  "dodgerblue": 0xFF1E90FF,
  "firebrick": 0xFFB22222,
  "floralwhite": 0xFFFFFAF0,
  "forestgreen": 0xFF228B22,
  "fuchsia": 0xFFFF00FF,
  "gainsboro": 0xFFDCDCDC,
  "ghostwhite": 0xFFF8F8FF,
  "gold": 0xFFFFD700,
  "goldenrod": 0xFFDAA520,
  "gray": 0xFF808080,
  "green": 0xFF008000,
  "greenyellow": 0xFFADFF2F,
  "grey": 0xFF808080,
  "honeydew": 0xFFF0FFF0,
  "hotpink": 0xFFFF69B4,
  "indianred": 0xFFCD5C5C,
  "indigo": 0xFF4B0082,
  "ivory": 0xFFFFFFF0,
  "khaki": 0xFFF0E68C,
  "lavender": 0xFFE6E6FA,
  "lavenderblush": 0xFFFFF0F5,
  "lawngreen": 0xFF7CFC00,
  "lemonchiffon": 0xFFFFFACD,
  "lightblue": 0xFFADD8E6,
  "lightcoral": 0xFFF08080,
  "lightcyan": 0xFFE0FFFF,
  "lightgoldenrodyellow": 0xFFFAFAD2,
  "lightgray": 0xFFD3D3D3,
  "lightgreen": 0xFF90EE90,
  "lightgrey": 0xFFD3D3D3,
  "lightpink": 0xFFFFB6C1,
  "lightsalmon": 0xFFFFA07A,
  "lightseagreen": 0xFF20B2AA,
  "lightskyblue": 0xFF87CEFA,
  "lightslategray": 0xFF778899,
  "lightslategrey": 0xFF778899,
  "lightsteelblue": 0xFFB0C4DE,
  "lightyellow": 0xFFFFFFE0,
  "lime": 0xFF00FF00,
  "limegreen": 0xFF32CD32,
  "linen": 0xFFFAF0E6,
  "magenta": 0xFFFF00FF,
  "maroon": 0xFF800000,
  "mediumaquamarine": 0xFF66CDAA,
  "mediumblue": 0xFF0000CD,
  "mediumorchid": 0xFFBA55D3,
  "mediumpurple": 0xFF9370DB,
  "mediumseagreen": 0xFF3CB371,
  "mediumslateblue": 0xFF7B68EE,
  "mediumspringgreen": 0xFF00FA9A,
  "mediumturquoise": 0xFF48D1CC,
  "mediumvioletred": 0xFFC71585,
  "midnightblue": 0xFF191970,
  "mintcream": 0xFFF5FFFA,
  "mistyrose": 0xFFFFE4E1,
  "moccasin": 0xFFFFE4B5,
  "navajowhite": 0xFFFFDEAD,
  "navy": 0xFF000080,
  "oldlace": 0xFFFDF5E6,
  "olive": 0xFF808000,
  "olivedrab": 0xFF6B8E23,
  "orange": 0xFFFFA500,
  "orangered": 0xFFFF4500,
  "orchid": 0xFFDA70D6,
  "palegoldenrod": 0xFFEEE8AA,
  "palegreen": 0xFF98FB98,
  "paleturquoise": 0xFFAFEEEE,
  "palevioletred": 0xFFDB7093,
  "papayawhip": 0xFFFFEFD5,
  "peachpuff": 0xFFFFDAB9,
  "peru": 0xFFCD853F,
  "pink": 0xFFFFC0CB,
  "plum": 0xFFDDA0DD,
  "powderblue": 0xFFB0E0E6,
  "purple": 0xFF800080,
  "red": 0xFFFF0000,
  "rosybrown": 0xFFBC8F8F,
  "royalblue": 0xFF4169E1,
  "saddlebrown": 0xFF8B4513,
  "salmon": 0xFFFA8072,
  "sandybrown": 0xFFF4A460,
  "seagreen": 0xFF2E8B57,
  "seashell": 0xFFFFF5EE,
  "sienna": 0xFFA0522D,
  "silver": 0xFFC0C0C0,
  "skyblue": 0xFF87CEEB,
  "slateblue": 0xFF6A5ACD,
  "slategray": 0xFF708090,
  "slategrey": 0xFF708090,
  "snow": 0xFFFFFAFA,
  "springgreen": 0xFF00FF7F,
  "steelblue": 0xFF4682B4,
  "tan": 0xFFD2B48C,
  "teal": 0xFF008080,
  "thistle": 0xFFD8BFD8,
  "tomato": 0xFFFF6347,
  "turquoise": 0xFF40E0D0,
  "violet": 0xFFEE82EE,
  "wheat": 0xFFF5DEB3,
  "white": 0xFFFFFFFF,
  "whitesmoke": 0xFFF5F5F5,
  "yellow": 0xFFFFFF00,
  "yellowgreen": 0xFF9ACD32,
};

final colorValueToName = Map<int, String>.unmodifiable(
    colorNameToValue.map((key, value) => MapEntry(value, key)));


lib/color_from.dart
import 'dart:math';

import 'package:color_tools/color_tabels.dart';
import 'package:flutter/painting.dart';

extension ColorFrom on Color {
  /// Get color from a int value without alpha
  ///
  /// ```dart
  /// ColorFrom.noAlpha(0xAABBCC) == Color(0xFFAABBCC)
  /// ```
  static Color noAlpha(int value) {
    return Color(0xff000000 | value);
  }

  /// Get color from a color name
  ///
  /// ```dart
  /// ColorFrom.fromName('red') == Color(0xFFFF0000)
  /// ```
  static Color? name(String name) {
    final data = colorNameToValue[name.toLowerCase()];
    return data != null ? Color(data) : null;
  }

  /// Get color a from a color code
  ///
  /// ```dart
  /// ColorFrom.code('#AABBCC') == Color(0xFFAABBCC)
  /// ColorFrom.code('#11AABBCC') == Color(0x11AABBCC)
  /// ```
  static Color code(String string) {
    assert(string.startsWith('#'), "hash string must start with #");
    final numbers = string.substring(1);
    final data = int.parse(numbers, radix: 16);
    if (numbers.length == 6) return Color(0xff000000 | data);
    return Color(data);
  }

  /// Try to get a color by name or code
  ///
  /// ```dart
  /// ColorFrom.string('red') == Color(0xFFFF0000)
  /// ColorFrom.string('#FF0000') == Color(0xFFFF0000)
  /// ```
  static Color? string(String string) {
    if (string.startsWith("#")) return ColorFrom.code(string);
    return ColorFrom.name(string);
  }

  /// Covert a object(hash) to a color
  ///
  /// The Color is deterministic and fully defind by the hash of the [object].
  /// The limits can be used to set the desired range of the output colors.
  /// By default all colors are possible
  ///
  ///
  /// [hueUpperLimit] and [hueLowerLimit] set the range of possible hues the
  /// value range is 0 to 360
  ///
  /// [saturationUpperLimit] and [saturationLowerLimit] set the range of
  /// possible saturations the value range is 0 to 1
  ///
  /// [valueUpperLimit] and [valueUpperLimit] set the range of
  /// possible values the value range is 0 to 1
  ///
  /// ```dart
  /// ColorFrom.object('some object')
  /// // it is deterministic!
  /// ColorFrom.object('a') == ColorFrom.object('a')
  /// ```
  static Color object(
    Object object, {
    final double hueUpperLimit = 360,
    final double hueLowerLimit = 0,
    final double saturationUpperLimit = 1,
    final double saturationLowerLimit = 0,
    final double valueUpperLimit = 1,
    final double valueLowerLimit = 0,
  }) {
    assert(hueUpperLimit <= 360);
    assert(hueUpperLimit >= 0);
    assert(hueLowerLimit <= 360);
    assert(hueLowerLimit >= 0);

    assert(saturationUpperLimit <= 1);
    assert(saturationUpperLimit >= 0);
    assert(saturationLowerLimit <= 1);
    assert(saturationLowerLimit >= 0);

    assert(valueUpperLimit <= 1);
    assert(valueUpperLimit >= 0);
    assert(valueLowerLimit <= 1);
    assert(valueLowerLimit >= 0);

    final random = Random(object.hashCode);
    final hue =
        random.nextDouble() * (hueUpperLimit - hueLowerLimit) + hueLowerLimit;
    final saturation =
        random.nextDouble() * (saturationUpperLimit - saturationLowerLimit) +
            saturationLowerLimit;
    final value = random.nextDouble() * (valueUpperLimit - valueLowerLimit) +
        valueLowerLimit;
    return HSVColor.fromAHSV(1, hue, saturation, value).toColor();
  }
}
